ggml/src/ggml-sycl/set_rows.cpp

```diff
@@ -0,0 +1,143 @@
+    const float * src_f = (const float *) src;
+    sycl::half * dst_h = (sycl::half *) dst;
+    *dst_h = sycl::vec<float, 1>(*src_f).convert<sycl::half, sycl::rounding_mode::automatic>()[0];
+}
+
+template<set_rows_kernel_t set_rows_1>
+static void k_set_rows(
+        const char * __restrict__ src0, const int64_t * __restrict__ src1, char * __restrict__ dst,
+        const int64_t ne00, const int64_t ne01, const int64_t ne11, const int64_t ne12,
+        const size_t nb01, const size_t nb02, const size_t nb03,
+        const size_t nb10, const size_t nb11, const size_t nb12,
+        const size_t nb1, const size_t nb2, const size_t nb3,
+        const size_t src_type_size, const size_t dst_type_size,
+        const sycl::nd_item<3> & item_ct1) {
+
+    const int i03 = item_ct1.get_group(0);
+    const int i02 = item_ct1.get_group(1);
+    const int i01 = item_ct1.get_group(2) * item_ct1.get_local_range(1) + item_ct1.get_local_id(1);  // Row index
+
+    if (i01 >= ne01) {
+        return;
+    }
+
+    const int i12 = i03 % ne12;
+    const int i11 = i02 % ne11;
+    const int i10 = i01;
+
+    const int64_t dst_row = *(const int64_t *)((const char *)src1 + i10*nb10 + i11*nb11 + i12*nb12);
+
+    const char * src0_row = src0 + i01*nb01 + i02*nb02 + i03*nb03;
+    char * dst_row_ptr    = dst + dst_row*nb1 + i02*nb2 + i03*nb3;
+    // Optimize for same-type operations: use collective memory copy
+    if (src_type_size == dst_type_size) {
+        // All threads in the work-group cooperatively copy the row
+        const size_t row_bytes = ne00 * src_type_size;
+        // Each thread copies a chunk of the row
+        for (size_t byte_idx = item_ct1.get_local_id(0); byte_idx < row_bytes; byte_idx += item_ct1.get_local_range(0)) {
+            dst_row_ptr[byte_idx] = src0_row[byte_idx];
+        }
+    } else {
+        // Type conversion required, use element-wise approach
+        for (int col = item_ct1.get_local_id(0); col < ne00; col += item_ct1.get_local_range(0)) {
+            const char * src_elem = src0_row + col * src_type_size;
+            char * dst_elem       = dst_row_ptr + col * dst_type_size;
+            set_rows_1(src_elem, dst_elem);
+        }
+    }
+}
+
+template<set_rows_kernel_t set_rows_1>
+static void set_rows_sycl(
+        const char * src0_d, const int64_t * src1_d, char * dst_d,
+        const int64_t ne00, const int64_t ne01, const int64_t ne02, const int64_t ne03,
+        const int64_t ne11, const int64_t ne12, const size_t nb01, const size_t nb02, const size_t nb03,
+        const size_t nb10, const size_t nb11, const size_t nb12,
+        const size_t nb1, const size_t nb2, const size_t nb3,
+        const size_t src_type_size, const size_t dst_type_size,
+        queue_ptr stream) {
+
+    const int max_threads_per_row = 128; // KEEPING 128 for now
+    const int threads_per_row     = std::min((int)ne00, max_threads_per_row);
+
+    const int max_threads_per_block = 128;
```

Comment:
Shouldn't this be 256?

Reply:
I tried with 256 initially since CUDA used it. Reverted!

Reply:
Yes. A value of 64 instead of 256 so far gives the best performance on my A750.

Reply:
I think it would be helpful if you are able to show us the results on a B580 or LL iGPU.